src/plugins/path/path.c

```diff
@@ -65,6 +100,149 @@ static int validateKey (Key * key, Key * parentKey)
 	return 1;
 }
 
+/**
+ * This method validates the file permission for a certain user
+ * The method assumes that the path exists and only validates permission
+ * @param key The key containting all metadata
+ * @param parentKey The parentKey which is used for error writing
+ * @return 0 if success or -1 for failure
+ */
+static int validatePermission (Key * key, Key * parentKey)
+{
+	uid_t currentUID = geteuid ();
+	gid_t currentGID = getegid ();
+
+	const Key * userMeta = keyGetMeta (key, "check/permission/user");
+	const Key * userTypes = keyGetMeta (key, "check/permission/types");
+
+	// ***** central variables *******
+	const char * validPath = keyString (key);
+	const char * name = keyString (userMeta);
+	const char * modes = keyString (userTypes);
+	// ****************************
+
+	struct passwd * p;
+
+	// Changing to specified user. Can only be done when executing user is root user
+	if (userMeta)
+	{
+		p = getpwnam (name);
+		// Check if user exists
+		if (p == NULL)
+		{
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_USER_NOT_FOUND, parentKey,
+					    "Could not find user \"%s\" for key \"%s\". "
+					    "Does the user exist?\"",
+					    name, keyName (key));
+			return -1;
+		}
+		name = p->pw_name;
+
+		// Check if I can change the UID as root
+		int err = seteuid ((int) p->pw_uid);
+		if (err < 0)
+		{
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CHANGE_USER_ERROR, parentKey,
+					    "Could not set euid of user \"%s\" for key \"%s\"."
+					    " Are you running kdb as root?\"",
+					    name, keyName (key));
+			return -1;
+		}
+	}
+	else
+	{
+		uid_t uid = geteuid ();
+		p = getpwuid (uid);
+		name = p->pw_name;
+		if (uid != 0)
+		{
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CHANGE_USER_ERROR, parentKey,
+					    "To check permissions for %s I need to be the root user."
+					    " Are you running kdb as root?\"",
+					    keyName (key));
+			return -1;
+		}
+	}
+
+	// The following code changes the egid if a group from a user matches the filegroup
+	// TODO: Whats a good default value for ngroups
```

Comment:
From the manpage for `getgrouplist(3)`:
> If the user is a member of more than *ngroups groups, then getgrouplist() returns -1.  In this case, the value returned in *ngroups can be used to resize the buffer passed to a further call getgrouplist().

So I think the idea was to try with an ngroups value of lets say 16. Check if getgrouplist() returns -1. If so, realloc to the size of the new ngroups value. Call getgrouplist() again. If it fails now, report an error (not sure if this is even possible).

elektraVFormat does something similar with `snprintf`, which always returns the minimum buffer size needed to fit the whole result.

Reply:
I implemented a different approach now, please take a look @kodebach if this is actually a decent way of doing such thing

Reply:
Your approach works but is acutal much more complicated than it needs to be.

You should need at most 2 calls to `getgrouplist`, because a failed call returns all the information needed to make a successful call (unless the number of groups changes between calls, if this is even possible).

This much simpler approach works as well:
```c
int ngroups = 0;
getgrouplist (p->pw_name, (int) p->pw_gid, (gid_t[]){}, &ngroups);
// call to getgrouplist fails because at least one group (p->pw_gid) is returned
// therefore ngroups now contains the actual number of groups for the user
gid_t *groups = (gid_t *)elektraMalloc(ngroups * sizeof(gid_t));
if (getgrouplist (p->pw_name, (int) p->pw_gid, groups, &ngroups) < 0)
{
    // TODO: error
}

```